### PR TITLE
add more YaLafi arguments to user preferences

### PR DESCRIPTION
**Pull request by torik42**
I just came across LanguageTool, YaLafi and this linter for atom. It’s great!

This pull request allows to set
- LanguageTool path
- n-gram data path
- LanguageTool server options
- disabled LanguageTool rules

from the user preferences.

### CODE REVIEW

**Comment by torik42** [comment]
I am wondering why the arguments are put together each time the linter runs. Is there a nice way to do that whenever the config changes?

**Comment by torik42** [comment]
I would like to add some documentation for the settings to the README and point to the corresponding YaLafi and LanguageTool documentation.

Also, it would be nice, if the LanguageTool (LT) server options (and the n-gram data path) only show up if one selects the localhost option. Do you know whether that is possible and desired?

Finally, I have one last concern: The LT server options only influence the server startup. Hence, changing the preferences will not influence the linter as long as the server is not killed. Unfortunately, for me `python -m yalafi.shell --server stop` does not work. Even if there is a possibility to stop the server, it is maybe not good to stop it in case the user started the server independent of atom. Hence, I would suggest to point this out in the README and maybe add an issue to think about this later.

What do you think?

**Comment by torik42** [comment]
I now added all the documentation I intended.

I also worked out how to make part of the settings expandable. But it then get’s a big heading which does not look good in my opinion. Hence I did not finished that and only reordered the settings.

File: README.md [hunk 1]
```diff
@@ -14,3 +14,18 @@ You also have to install [linter](https://atom.io/packages/linter):
 Since this is an interface to YaLafi it also has to be installed. The easiest way is to use `pip`:
 
 `pip install yalafi`
+
+## Using a local LanguageTool Server
+
+If you do not want to use the LanguageTool online server you also need to install Java and  LanguageTool.
+More information on how to install LanguageTool can be found [here](https://github.com/matze-dd/YaLafi#installation).
+
+Then you can set the 'LanguageTool path' in the settings.
```
Comment (mfbehrens99) [inline README.md:23]: or start the server manually

File: lib/linter-yalafi.js [hunk 2]
```diff
@@ -160,15 +172,24 @@ export default {
...
+        if (this.languageToolnGramDataPath !== "") {
+          languageToolServerOptions = '--languageModel ' + this.languageToolnGramDataPath + ' ' + this.languageToolServerOptions
+        }
         //TODO Add options
         const args = [
           '-m',
           'yalafi.shell',
           '--server=' + this.server,
           '--output=json',
+          "--lt-server-options='" + languageToolServerOptions + "'",
```
Comment (mfbehrens99) [inline lib/linter-yalafi.js:186]: I think you don't need the `"` but you should check it again

Comment (torik42) [reply lib/linter-yalafi.js:186]: I need to have `'` in the string itself. Thus I used `"`. Alternatively I could use `\'` inside the string if you like it better.

Comment (mfbehrens99) [reply lib/linter-yalafi.js]: Sorry my comment was incorrect. I wanted to say: I think you don't need the `'` however I am not sure how this input is used. I only know that `"` or `'` is not needed for the file path.

However, this is such a minor thing, so it is not that important.

**Comment by mfbehrens99** [comment]
Okay my comment about starting the LT server manually was confusing. I always have the LT server running since I use the Chrome extension and some other services that rely on LT. I only wanted to add something like:
Then you can set the 'LanguageTool path' in the settings if your LanguageTool server is not already running.

**Comment by mfbehrens99** [comment]
> I am wondering why the arguments are put together each time the linter runs. Is there a nice way to do that whenever the config changes?

I think this is a good idea, but it is not too important. I'll add an issue for that. Maybe we can enhance that in future

>Also, it would be nice, if the LanguageTool (LT) server options (and the n-gram data path) only show up if one selects the localhost option. Do you know whether that is possible and desired?

I researched that, but I couldn't find anything on that topic and I have never seen something like that in a package's settings. But If you find any way to do this, it would be great.

>Finally, I have one last concern: The LT server options only influence the server startup. Hence, changing the preferences will not influence the linter as long as the server is not killed. Unfortunately, for me python -m yalafi.shell --server stop does not work. Even if there is a possibility to stop the server, it is maybe not good to stop it in case the user started the server independent of atom. Hence, I would suggest to point this out in the README and maybe add an issue to think about this later.

Even if it would work It would make the startup time very long. Additionally, when a LanguageTool server is already running with different settings I am not sure how this situation should be handled. I think when people are using LT with many settings they are not only using it with `linter-yalafi` but also with other programs. In that case they probably have an autostart file in there system folder to start the server with all the settings the want and need.
In my opinion, `linter-yalafi` should only start the LT-server when no server is running.

File: lib/linter-yalafi.js (hunk 2 above)
Comment (torik42) [reply lib/linter-yalafi.js]: This is from the YaLafi README
>`--lt-server-options opts`
  Pass additional options when starting a local LT server.
  Syntax is as for --lt-options.

> `--lt-options opts`
  Pass additional options to LT, given as single string in argument 'opts'.
  The first character of 'opts' will be skipped and must not be '-'.
  Example:
  `--lt-options '~--languagemodel ../Ngrams --disablecategories PUNCTUATION'`.
  Some options are included into HTML requests to an LT server, see script
  variable 'lt\_option\_map'.

I need the `'` such that YaLafi passes the string, i.e. the options, to the LanguageTool server. Somehow, when I tested it, the first character wasn’t removed. Thats why there is no `~` in my code.

**Comment by torik42** [comment]
> Okay my comment about starting the LT server manually was confusing. I always have the LT server running since I use the Chrome extension and some other services that rely on LT. I only wanted to add something like:
> Then you can set the 'LanguageTool path' in the settings if your LanguageTool server is not already running.

To me it seems that I cannot call YaLafi with `--lt-server my` if the path is not specified or the variable 'ltdirectory' is properly set in `yalafi/shell/shell.py`. So even if you have the server running, you may have set the latter properly, right?

Hence, even if the local LanguageTool server is already running (at port 8081) and you only use YaLafi within atom without worrying about the config in YaLafi, you would need to set the path, don’t you?


> > Also, it would be nice, if the LanguageTool (LT) server options (and the n-gram data path) only show up if one selects the localhost option. Do you know whether that is possible and desired?
>
> I researched that, but I couldn't find anything on that topic and I have never seen something like that in a package's settings. But If you find any way to do this, it would be great.

As I wrote above, it is possible to make part of the settings collapsable (as e.g. in the 'ide-python' package). It will not disappear completely (which I like) and there is some rather large heading (which I don’t like). But it’s rather easy to implement, but not necessary.

> Even if it would work It would make the startup time very long. Additionally, when a LanguageTool server is already running with different settings I am not sure how this situation should be handled. I think when people are using LT with many settings they are not only using it with `linter-yalafi` but also with other programs. In that case they probably have an autostart file in there system folder to start the server with all the settings the want and need.
> In my opinion, `linter-yalafi` should only start the LT-server when no server is running.

Yes, I totally agree. That’s why I wrote into the README that the user has to kill the server manually.

**Comment by mfbehrens99** [comment]
>To me it seems that I cannot call YaLafi with --lt-server my if the path is not specified or the variable 'ltdirectory' is properly set in yalafi/shell/shell.py. So even if you have the server running, you may have set the latter properly, right?

Before you joined the development I intended the package to require the server to be started when you want to use the linter. I didn't have to set any variables or paths for it to work when the server was already started.

I still like the feature that the server is automatically launched when it has not been started yet.

>As I wrote above, it is possible to make part of the settings collapsable (as e.g. in the 'ide-python' package). It will not disappear completely (which I like) and there is some rather large heading (which I don’t like). But it’s rather easy to implement, but not necessary.

As I already wrote in #8 I think this would be good to keep the settings tidy and easy to use (for not devs).

#9 is an option which I think would be really nice. If you want you can add this option two.

**Comment by torik42** [comment]
I now added an option to enter arbitrary YaLafi arguments. We postponed the grouping of the preferences. Is there anything else to do?

**Pull request merged**